storage_service/locations/api/beta/remple/openapi.py

```diff
@@ -0,0 +1,1573 @@
+            ]),
+            'order_by_subattribute': OrderedDict([
+                ('in', 'query'),
+                ('name', 'order_by_subattribute'),
+                ('schema', {'type': 'string'}),
+                ('required', False),
+                ('description', 'Attribute of the related attribute'
+                                ' order_by_attribute of the resource'
+                                ' that view results should be'
+                                ' ordered by.'),
+            ]),
+            'order_by_direction': OrderedDict([
+                ('in', 'query'),
+                ('name', 'order_by_direction'),
+                ('schema', OrderedDict([
+                    ('type', 'string'),
+                    ('enum', [obd for obd in QueryBuilder.order_by_directions
+                              if obd]),
+                ])),
+                ('required', False),
+                ('description', 'The direction of the ordering; omitting this'
+                                ' parameter means ascending direction.'),
+            ])
+        }
+
+    def _get_error_schema(self):
+        return OrderedDict([
+            ('type', 'object'),
+            ('properties', OrderedDict([
+                ('error', OrderedDict([
+                    ('type', 'string'),
+                ])),
+            ])),
+            ('required', ['error']),
+        ])
+
+    def _get_custom_endpoints(self, resource_cls):
+        """Return the custom endpoints defined on ``resource_cls``.
+        """
+        ret = []
+        for attr_name in dir(resource_cls):
+            if attr_name.startswith('_'):
+                continue
+            custom_endpoint = getattr(resource_cls, attr_name)
+            if isinstance(custom_endpoint, CustomEndPoint):
+                ret.append(custom_endpoint)
+        return ret
+
+    def _get_cust_endpt_schemata(self, resource_name, resource_cfg):
+        """Return an interable of custom endpoint schema names and schemata for
+        custom endpoints defined on the resource class.
+        """
+        ret = []
+        for custom_endpoint in self._get_custom_endpoints(
+                resource_cfg['resource_cls']):
+            for schema_name, schema in custom_endpoint.schemata.items():
+                ret.append((schema_name, schema))
+        return ret
+
+    def _get_schemas(self):
+        """Return an OpenAPI ``schemas`` OrderedDict.
+
+        It contains a "View", a "Create", and an "Update" schema for each
+        resource in ``self.resources``, e.g., ``LocationView``,
+        ``LocationCreate``, and ``LocationUpdate``. The create and update
+        schemata are only included if the resources is not read-only. The view
+        schema is constructed by introspecting the Django model, while the
+        create and update schemata are constructed by introspecting the relevant
+        Formencode schemata attached as class attributes on the resource class.
+        """
```

Comment:
TODO: it seems that OpenAPI's readOnly and writeOnly boolean property attributes could be used here. See https://swagger.io/docs/specification/data-models/data-types/, in particular the Read-Only and Write-Only Properties section.